### DISCUSSION TITLE
.mod files during build

### DISCUSSION THREAD

**Discussion by ejovo13** [fortran-lang/fpm#471] [General]
Hello all,

I've been interested in Fortran recently and this repository seems like a cool place to be able to contribute to. I've been playing around with fpm to get a feel for how it works and I had a question about the creation of mod files. Using the default project that is created automatically, I noticed that after a call to fpm build the following command is printed to the screen:
` + gfortran -c ././src/test2.f90 -Wall -Wextra -Wimplicit-interface -fPIC -fmax-errors=1 -g -fcheck=bounds -fcheck=array-temps -fbacktrace -fcoarray=single -J build/gfortran_2A42023B310FA28D/test2 -I build/gfortran_2A42023B310FA28D/test2 -o build/gfortran_2A42023B310FA28D/test2/src_test2.f90.o`
The switch -J indicates that .mod files should be outputted to the  build/gfortran_2A42023B310FA28D/test2 directory, but in fact there are no .mod files. I was wondering, then, how the main.f90 (under app dir) is able to access the module's interface. Are the .mod files created, and then wiped after linking? When I copy and paste that command manually, the test2.mod file appears (as expected) in the -J directory. Another side question that I have is: What is the purpose of the .digest files?

**Comment by awvwgk** [comment]
The `.digest` files contain the hash of the respective source files which fpm uses to decide whether or not to recompile object files.

**Reply by ejovo13** [reply]
Thank you

**Comment by LKedward** [comment]
Welcome Evan @ejovo13 and thanks for your interest in *fpm*. I'm unsure why in your case there are no `.mod` files - you are correct to expect `.mod` files in that output directory and fpm does not delete them after building. Please see my video attached below.

Great to hear that you're interested in contributing! If you haven't already, you can have a read through the developer documentation [here](https://fpm.fortran-lang.org/) to learn more about how fpm works inside. Do feel free to ask us more questions, we're happy to help you get to know fpm and start contributing!

https://user-images.githubusercontent.com/26024234/116779543-bed9c780-aa6e-11eb-98f5-6d0732875166.mp4

**Reply by ejovo13** [reply]
Thank you. Of course, now I am unable to reproduce the "issue" and the `.mod` files are showing up as expected. I'm using vscode as my primary editor and I know that it can be a little finicky with `.mod` files, so maybe something was going wrong there. When I run `fpm build` from a separate terminal instance there are no problems. 

Currently reading through the documentation and packaging guide. I saw in the issues tab under "good first issue" there are some opportunities to integrate some known packages to work with `fpm`. Up until now I've only worked on solo projects, so I don't _exactly_ know the proper way to go about doing this. For example. theoretically, how would I go about getting stdlib to work with fpm? Just slap a `fpm.toml` file in the parent directory? Surely it's more complicated then that and I was wondering if you could provide a brief outline of what steps to take to get an existing package (that I'm not the author of) compatible with `fpm`.

**Reply by LKedward** [reply]
Ah I see - I haven't run fpm from vscode so I'm not sure what nuances may arise there.

Yes converting existing packages is a great place to start, but it does involve more than adding an `fpm.toml` file. The packaging guide should cover most of it but here's a rough outline:

1. Determine if the package can currently be built with fpm; fpm can't yet build packages needing third-party preprocessors (like fypp needed in stdlib) or packages with complex build configurations such as optional/conditional features. (So best choose a simple project to start off with)

2. Separate out the package sources into library sources, app sources and test sources, and place these in the default locations for fpm (`src/`, `app/`, `test/`)

3. Add an `fpm.toml` file and add any [`link`](https://github.com/fortran-lang/fpm/blob/master/manifest-reference.md#link-external-libraries) entries and fpm [dependencies](https://github.com/fortran-lang/fpm/blob/master/manifest-reference.md#specifying-dependencies) if used

4. Try to build the package and run the tests/programs

It's worth checking the existing [fpm packages](https://fortran-lang.org/packages/fpm) as well as the [example packages](https://github.com/fortran-lang/fpm/tree/master/example_packages) if you're unsure of what fpm can handle. Finally, here's a package that I recently restructured for fpm: https://github.com/kookma/ogpf

**Reply by ejovo13** [reply]
Thank you! I'll keep educating myself and will hopefully be in a position to help out in some small part over the coming months!
